pkg/bindinfo/tests/bind_usage_info_test.go

```diff
@@ -0,0 +1,126 @@
+
+import (
+	"testing"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/bindinfo"
+	"github.com/pingcap/tidb/pkg/testkit"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBindUsageInfo(t *testing.T) {
+	checklist := []string{
+		"5ce1df6eadf8b24222668b1bd2e995b72d4c88e6fe9340d8b13e834703e28c32",
+		"5d3975ef2160c1e0517353798dac90a9914095d82c025e7cd97bd55aeb804798",
+		"9d3995845aef70ba086d347f38a4e14c9705e966f7c5793b9fa92194bca2bbef",
+		"aa3c510b94b9d680f729252ca88415794c8a4f52172c5f9e06c27bee57d08329",
+	}
+	bindinfo.UpdateBindingUsageInfoBatchSize = 2
+	bindinfo.MaxWriteInterval = 100 * time.Microsecond
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	bindingHandle := dom.BindingHandle()
+	tk := testkit.NewTestKit(t, store)
+
+	tk.MustExec(`use test`)
+	tk.MustExec(`set @@tidb_opt_enable_fuzzy_binding=1`)
+	tk.MustExec("create table t1(a int, b int, c int, key idx_b(b), key idx_c(c))")
+	tk.MustExec("create table t2(a int, b int, c int, key idx_b(b), key idx_c(c))")
+	tk.MustExec("create table t3(a int, b int, c int, key idx_b(b), key idx_c(c))")
+	tk.MustExec("create table t4(a int, b int, c int, key idx_b(b), key idx_c(c))")
+	tk.MustExec("create table t5(a int, b int, c int, key idx_b(b), key idx_c(c))")
+
+	tk.MustExec("prepare stmt1 from 'delete from t1 where b = 1 and c > 1';")
+	tk.MustExec("prepare stmt2 from 'delete t1, t2 from t1 inner join t2 on t1.b = t2.b';")
+	tk.MustExec("prepare stmt3 from 'update t1 set a = 1 where b = 1 and c > 1';")
+	tk.MustExec("execute stmt1;")
+	tk.MustExec("create global binding for delete from t1 where b = 1 and c > 1 using delete /*+ use_index(t1,idx_c) */ from t1 where b = 1 and c > 1")
+	tk.MustExec("create global binding for delete t1, t2 from t1 inner join t2 on t1.b = t2.b using delete /*+ inl_join(t1) */ t1, t2 from t1 inner join t2 on t1.b = t2.b")
+	tk.MustExec("create global binding for update t1 set a = 1 where b = 1 and c > 1 using update /*+ use_index(t1,idx_c) */ t1 set a = 1 where b = 1 and c > 1")
+	// corss database binding
+	tk.MustExec(`create global binding using select /*+ leading(t1, t2, t3, t4, t5) */ * from *.t1, *.t2, *.t3, *.t4, *.t5`)
+	tk.MustExec("select * from t1, t2, t3, t4, t5")
+	tk.MustExec("execute stmt1;")
+	origin := tk.MustQuery(`select sql_digest,last_used_date from mysql.bind_info where original_sql != 'builtin_pseudo_sql_for_bind_lock' order by sql_digest`)
+	origin.Check(testkit.Rows(
+		"5ce1df6eadf8b24222668b1bd2e995b72d4c88e6fe9340d8b13e834703e28c32 <nil>",
+		"5d3975ef2160c1e0517353798dac90a9914095d82c025e7cd97bd55aeb804798 <nil>",
+		"9d3995845aef70ba086d347f38a4e14c9705e966f7c5793b9fa92194bca2bbef <nil>",
+		"aa3c510b94b9d680f729252ca88415794c8a4f52172c5f9e06c27bee57d08329 <nil>"))
+	time.Sleep(50 * time.Microsecond)
+	require.NoError(t, bindingHandle.UpdateBindingUsageInfoToStorage())
+	result := tk.MustQuery(`select sql_digest,last_used_date from mysql.bind_info where original_sql != 'builtin_pseudo_sql_for_bind_lock' order by sql_digest`)
```

Comment:
Again, I think we should use the `bindinfo.BuiltinPseudoSQL4BindLock` constant, and not use the literal hardcoded string `'builtin_pseudo_sql_for_bind_lock'`:

```suggestion
	result := tk.MustQuery(fmt.Sprintf(`select sql_digest,last_used_date from mysql.bind_info where original_sql != '%s' order by sql_digest`, bindinfo.BuiltinPseudoSQL4BindLock))
```

Please make sure all uses of `'builtin_pseudo_sql_for_bind_lock'` in this file are changed.

Notice that the literal string `"builtin_pseudo_sql_for_bind_lock"` is never used outside of pkg/bindinfo/binding_handle.go. Also, using the constant instead of the literal string is how `TestRemoveDuplicatedPseudoBinding` in pkg/bindinfo/binding_operator_test.go works.

For a magic value/constant like the name `"builtin_pseudo_sql_for_bind_lock"`, it is generally better to refer to its symbolic constant rather than its literal value, because it means the value is only defined in one place. If the value needs to change, it only needs to be updated in one place, rather than in every place where it is used.

Reply:
Updated